pkg/volume/flocker/client.go

```diff
@@ -0,0 +1,391 @@
+}
+
+// post performs a post request with the indicated payload
+func (c flockerClient) post(url string, payload interface{}) (*http.Response, error) {
+	return c.request("POST", url, payload)
+}
+
+// get performs a get request
+func (c flockerClient) get(url string) (*http.Response, error) {
+	return c.request("GET", url, nil)
+}
+
+// getURL returns a full URI to the control service
+func (c flockerClient) getURL(path string) string {
+	return fmt.Sprintf("%s://%s:%d/%s/%s", c.schema, c.host, c.port, c.version, path)
+}
+
+type configurationPayload struct {
+	Primary     string          `json:"primary"`
+	DatasetID   string          `json:"dataset_id,omitempty"`
+	MaximumSize json.Number     `json:"maximum_size,omitempty"`
+	Metadata    metadataPayload `json:"metadata,omitempty"`
+}
+
+type metadataPayload struct {
+	Name string `json:"name,omitempty"`
+}
+
+type datasetState struct {
+	Path      string `json:"path"`
+	DatasetID string `json:"dataset_id"`
+}
+
+type datasetStatePayload struct {
+	*datasetState
+}
+
+type nodeStatePayload struct {
+	UUID string `json:"uuid"`
+	Host string `json:"host"`
+}
+
+// findIDInConfigurationsPayload returns the datasetID if it was found in the
+// configurations payload, otherwise it will return an error.
+func (c flockerClient) findIDInConfigurationsPayload(body io.ReadCloser, name string) (datasetID string, err error) {
+	var configurations []configurationPayload
+	if err = json.NewDecoder(body).Decode(&configurations); err == nil {
+		for _, r := range configurations {
+			if r.Metadata.Name == name {
+				return r.DatasetID, nil
+			}
+		}
+		return "", errConfigurationNotFound
+	}
+	return "", err
+}
+
+// findPathInDatasetStatePayload returns the path of the given datasetID if it
+// was found in the state payload. In case the path is not found it returns an
+// error.
+func (c flockerClient) findPathInDatasetStatePayload(body io.ReadCloser, datasetID string) (path string, err error) {
+	var states []datasetStatePayload
+	if err = json.NewDecoder(body).Decode(&states); err == nil {
+		for _, s := range states {
+			if s.DatasetID == datasetID {
+				return s.Path, nil
+			}
+		}
+		return "", errStateNotFound
+	}
+	return "", err
+}
+
+// findUUIDInNodeStatePayload returns the UUID for the given host.
+func (c flockerClient) findUUIDInNodeStatePayload(body io.ReadCloser) (uuid string, err error) {
+	var states []nodeStatePayload
+	if err = json.NewDecoder(body).Decode(&states); err == nil {
+		for _, s := range states {
+			if s.Host == c.host {
+				return s.UUID, nil
+			}
+		}
+		return "", errStateNotFound
+	}
+	return "", err
+}
+
+// getDatasetState performs a get request to get the state of the given datasetID, if
+// something goes wrong or the datasetID was not found it returns an error.
+func (c flockerClient) getDatasetState(datasetID string) (*datasetState, error) {
+	resp, err := c.get(c.getURL("state/datasets"))
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
```

Comment:
it appears caller of `getDatasetState` already remembers to close the connection.